rpc: Return accurate results for scanblocks

aureleoules opened the pull request:
Implements #26322.
Adds a `filter_false_positives` mode to `scanblocks` to accurately verify results from blockfilters.

If the option is enabled, pre-results given by blockfilters will be filtered out again by checking vouts and vins of all transactions of the relevant blocks against the given descriptors.

### Master
```bash
./src/bitcoin-cli -testnet -named scanblocks action=start scanobjects='["addr(tb1qcxf2gv93c26s6mqz7y6etpqdf70zmn67dualgr)"]'
{
  "from_height": 0,
  "to_height": 2376055,
  "relevant_blocks": [
    "000000000001bc35077dec4104e0ab1f667ae27059bd907f9a8fac55c802ae36",
    "00000000000120a9c50542d73248fb7c37640c252850f0cf273134ad9febaf61",
    "0000000000000082f7af3835da8b6146b0bfb243b8842f09c495fa1e74d454ed",
    "0000000000000094c32651728193bfbe91f6789683b8d6ac6ae2d22ebd3cb5d3"
  ]
}
```

### PR (without `filter_false_positives` mode)
Same as master
```bash
./src/bitcoin-cli -testnet -named scanblocks action=start scanobjects='["addr(tb1qcxf2gv93c26s6mqz7y6etpqdf70zmn67dualgr)"]' filter_false_positives=false
{
  "from_height": 0,
  "to_height": 2376055,
  "relevant_blocks": [
    "000000000001bc35077dec4104e0ab1f667ae27059bd907f9a8fac55c802ae36",
    "00000000000120a9c50542d73248fb7c37640c252850f0cf273134ad9febaf61",
    "0000000000000082f7af3835da8b6146b0bfb243b8842f09c495fa1e74d454ed",
    "0000000000000094c32651728193bfbe91f6789683b8d6ac6ae2d22ebd3cb5d3"
  ]
}
```

### PR (with `filter_false_positives` mode)
```bash
./src/bitcoin-cli -testnet -named scanblocks action=start scanobjects='["addr(tb1qcxf2gv93c26s6mqz7y6etpqdf70zmn67dualgr)"]' filter_false_positives=true
{
  "from_height": 0,
  "to_height": 2376058,
  "relevant_blocks": [
    "0000000000000082f7af3835da8b6146b0bfb243b8842f09c495fa1e74d454ed",
    "0000000000000094c32651728193bfbe91f6789683b8d6ac6ae2d22ebd3cb5d3"
  ]
}
```

Also adds a test to check that the blockhash of a transaction will be included in the `relevant_blocks` whether the `filter_false_positives` mode is enabled or not.

aureleoules committed:
rpc: Return accurate results for scanblocks

This makes use of undo data to accurately verify results
from blockfilters.

aureleoules commented:
I'm wondering whether the `accurate` mode should be enabled by default. If an application or a user uses the `scanblocks` RPC, they'll need to filter out false positives anyway, so maybe Bitcoin Core should do it?

w0xlt commented on src/rpc/blockchain.cpp:
```diff
@@ -2244,7 +2276,8 @@ static RPCHelpMan scanblocks()
             RPCArg{"stop_height", RPCArg::Type::NUM, RPCArg::DefaultHint{"chain tip"}, "Height to stop to scan"},
-            RPCArg{"filtertype", RPCArg::Type::STR, RPCArg::Default{BlockFilterTypeName(BlockFilterType::BASIC)}, "The type name of the filter"}
+            RPCArg{"filtertype", RPCArg::Type::STR, RPCArg::Default{BlockFilterTypeName(BlockFilterType::BASIC)}, "The type name of the filter"},
+            RPCArg{"accurate", RPCArg::Type::BOOL, RPCArg::Default{false}, "Enable accurate mode (requires -txindex). Inaccurate mode may result in false positives but is faster."},
```
Shouldn't it be optional ?

aureleoules replied:
It is optional.

w0xlt reviewed (commented):
Approach ACK.

w0xlt reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/26325/commits/ca139abad173b9142189f52e3dd1178684f81c0c

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -19,12 +19,14 @@
 #include <hash.h>
 #include <index/blockfilterindex.h>
 #include <index/coinstatsindex.h>
+#include <index/txindex.h>
```
Can't you read the scriptpubkey from the undo data?

theStack replied:
Isn't undo data only kept for recent blocks in the case of reorgs? Didn't check but I can hardly imagine that we store every scriptPubKey of a spent output twice forever.

aureleoules replied:
I didn't know about undo data, I pushed a version using it instead of `txindex`. Can revert if undo data is not reliable :/

theStack replied:
Seems my suspicion was wrong and we indeed keep undo data forever on a non-pruned node. Sorry for confusing!

andrewtoth replied:
> EDIT: I looked it up and undo data doesn't seem to ever be pruned so I think this is safe?

It does indeed get pruned at the same time as block files https://github.com/bitcoin/bitcoin/blob/master/src/node/blockstorage.cpp#L557-L565.

aureleoules replied:
> It does indeed get pruned at the same time as block files [master/src/node/blockstorage.cpp#L557-L565](https://github.com/bitcoin/bitcoin/blob/master/src/node/blockstorage.cpp?rgh-link-date=2022-10-19T19%3A54%3A57Z#L557-L565).

I meant that they are not pruned on non-pruned nodes but you are correct. I've added a check that throws if both pruning and filter_false_positives modes are enabled.

andrewtoth replied:
But you already added the `GetBlockChecked` and `GetUndoChecked` functions which check and throw if pruned. I don't think it's necessary to disable completely if pruned, since the blocks that are found might not have been pruned yet.

aureleoules replied:
Alright, reverted to 134caa7c25fee8ed16d0c390cf729f3033d06dc4.

theStack commented on src/rpc/blockchain.cpp:
```diff
@@ -2234,6 +2235,39 @@ class BlockFiltersScanReserver
+    }
+
+    CBlockUndo block_undo;
+    if(!UndoReadFromDisk(block_undo, pindex)) {
```
nit: spacing
```suggestion
    if (!UndoReadFromDisk(block_undo, pindex)) {
```

theStack reviewed (commented):
Concept ACK

> Also, is there a way to reproduce false positives? It would be nice to add a test but I haven't figured out how?

Discussed the same question this with @furszy just a few days ago talking about #26322, and came to the conclusion that the easiest way to find false-positives is to only use a single block with some random outputs, but increase the needle set (e.g. a ranged descriptor with a large range like 0-10000), until the RPC call returns a block. The found result can then be used in the test. Will tinker a bit around and see if I can come up with a commit that you can include (or in a follow-up).

aureleoules commented:
Thanks for the tips @theStack.
I tried creating a test as well. I managed to generate false positives using the following test, but it only happens once every few runs. I'll continue to work on it later, but here's my snippet in case you have ideas.


<details>
<summary>Test</summary>

```diff
diff --git a/test/functional/rpc_scanblocks.py b/test/functional/rpc_scanblocks.py
index 0ba5f12bd..e7d26a5ef 100755
--- a/test/functional/rpc_scanblocks.py
+++ b/test/functional/rpc_scanblocks.py
@@ -102,6 +102,42 @@ class ScanblocksTest(BitcoinTestFramework):
         # test invalid command
         assert_raises_rpc_error(-8, "Invalid command", node.scanblocks, "foobar")
 
+        # generate block with a lot of random outputs
+        print("Generate block with random transactions")
+        for i in range(500): # should be deterministic and give false positives
+            _, spk, _ = getnewdestination()
+            wallet.send_to(from_node=self.nodes[1], scriptPubKey=spk, amount=400)
+
+        blockhash = self.generate(self.nodes[1], 1)[0]
+        print(blockhash)
+
+        # scan the block with the ranged descriptor
+        height = node.getblockheader(blockhash)['height']
+        desc = f"pkh({parent_key}/*)"
+
+        # begin debug
+        blocks = node.scanblocks(
+            action="start",
+            scanobjects=[{"desc": desc, "range": [0, 200000]}],
+            start_height=height,
+            accurate=False)['relevant_blocks']
+
+        print(blocks)
+        # end debug
+
+        for v in [False, True]:
+            blocks = node.scanblocks(
+                action="start",
+                scanobjects=[{"desc": desc, "range": [0, 200000]}],
+                start_height=height,
+                accurate=v)['relevant_blocks']
+
+            if v:
+                assert_equal(len(blocks), 0)
+            else:
+                assert_equal(len(blocks), 1)
+                assert_equal(blocks[0], blockhash)
+
 
 if __name__ == '__main__':
     ScanblocksTest().main()
```

</details>

The only part that isn't deterministic is:
```python
for i in range(500): # should be deterministic and give false positives
	_, spk, _ = getnewdestination()
	wallet.send_to(from_node=self.nodes[1], scriptPubKey=spk, amount=400)
```

Maybe we could use `i` to generate deterministic scriptPubKeys and hope a certain arrangement always gives false positives? Not sure how reliable this is.

aureleoules commented:
I'm able to generate false positives consistently enough with this test by increasing `nblocks` but the test is very slow and still likely to cause intermittent failures so not worth adding it to the PR.

<details>
<summary>Details</summary>

```diff
diff --git a/test/functional/rpc_scanblocks.py b/test/functional/rpc_scanblocks.py
index 0ba5f12bd..05779973e 100755
--- a/test/functional/rpc_scanblocks.py
+++ b/test/functional/rpc_scanblocks.py
@@ -102,6 +102,34 @@ class ScanblocksTest(BitcoinTestFramework):
         # test invalid command
         assert_raises_rpc_error(-8, "Invalid command", node.scanblocks, "foobar")
 
+        # generate block with a lot of random outputs
+        print("Generate block with random transactions")
+        height = node.getblockcount() + 1
+        nblocks = 10
+        for i in range(nblocks):
+            for _ in range(550):
+                _, spk, _ = getnewdestination()
+                wallet.send_to(from_node=self.nodes[1], scriptPubKey=spk, amount=400)
+
+            print(f"Generate block {i}/{nblocks}")
+            self.generate(self.nodes[1], 1)[0]
+
+        # scan the block with the ranged descriptor
+        desc = f"pkh({parent_key}/*)"
+        print(desc)
+
+        for v in [False, True]:
+            blocks = node.scanblocks(
+                action="start",
+                scanobjects=[{"desc": desc, "range": [0, 200000]}],
+                start_height=height,
+                accurate=v)['relevant_blocks']
+
+            if v:
+                assert_equal(len(blocks), 0)
+            else:
+                assert len(blocks) > 1
+
 
 if __name__ == '__main__':
     ScanblocksTest().main()
```

</details>

andrewtoth commented:
I'm not sure `accurate` is very descriptive about what this feature does. Would `filter_false_positives` be a better name for this option?

> I'm wondering whether the accurate mode should be enabled by default. If an application or a user uses the scanblocks RPC, they'll need to filter out false positives anyway, so maybe Bitcoin Core should do it?

With this feature enabled Bitcoin Core will be reading from disk and scanning both the block and undo files for each block, which would then be read again when fetched and then scanned by the user. I don't think it would be worth this extra effort unless the user's network latency or scanning code is very slow so that a single false positive block getting to them would be slower than having Bitcoin Core do it for every block. So IMO it should be disabled by default.

andrewtoth commented on src/rpc/blockchain.cpp:
```diff
@@ -2234,6 +2235,39 @@ class BlockFiltersScanReserver
+{
+    CBlock block;
+    if (!ReadBlockFromDisk(block, pindex, chainparams.GetConsensus())) {
+        return false;
```
Here and below when reading the undo data, if the read fails then this returns false and the block is recorded as a false positive and won't be returned by `scanblocks`. These should instead throw a `JSONRPCError` because the results will not be correct.

andrewtoth commented:
I believe `scanblocks` is compatible with a pruned node as long as the block filters were generated before the blocks were pruned (correct me if I'm wrong please). If so, then this mode will fail silently for the pruned blocks and just not return them. But also more generally if a block read fails for any reason it should throw.

theStack commented:
> I'm able to generate false positives consistently enough with this test by increasing `nblocks` but the test is very slow and still likely to cause intermittent failures so not worth adding it to the PR.

I digged a bit deeper into the rabbit-hole and ended up opening #26341, adding a testing for a fixed false-positive for the genesis block and also verifying it with the ranged hashes described in BIP158. Feel free to base you PR on that (though I think the PR is already in a pretty good state and it's also fine if the test is add in a follow-up). With the newly introduced helpers, it would be possible to calculate false-positives for newly created blocks at runtime, but at least in my impression that takes too much time for a functional test -- about ~800k scriptPubKeys have to be tried out on average until one of them corresponds to a scriptPubKey already on the block.

aureleoules commented:
Thank you @andrewtoth for the review.
- Rebased this PR on top of #26341 as suggested by @theStack.
- Added a test case to filter false positives
- Added a check to not fetch block undo data for the genesis block

theStack commented on src/rpc/blockchain.cpp:
```diff
@@ -2234,6 +2235,40 @@ class BlockFiltersScanReserver
 
+static bool CheckBlockFilterMatches(const CBlockIndex* pindex, const GCSFilter::ElementSet& needles, const CChainParams& chainparams)
+{
+    CBlock block;
+    if (!ReadBlockFromDisk(block, pindex, chainparams.GetConsensus())) {
+        throw JSONRPCError(RPC_DATABASE_ERROR, "Cannot read block from disk");
+    }
+
+    CBlockUndo block_undo;
+    // Genesis block has no undo data
+    if (pindex->nHeight > 0 && !UndoReadFromDisk(block_undo, pindex)) {
+        throw JSONRPCError(RPC_DATABASE_ERROR, "Cannot read block undo data from disk");
+    }
```
Would be nice to use the helper functions `GetBlockChecked` and `GetUndoChecked` here to have consistent error messages and RPC return codes and avoid repetition. The only drawback right now is that `GetUndoChecked` doesn't include the exception for the genesis blocks yet, so the test would fail. Fixed by #19888 or #26369.

aureleoules replied:
Pushed suggestion https://github.com/bitcoin/bitcoin/compare/33793d67bff6519b6ebdd0b1229d03166c505672..50a145c39986dd7ed4a5667d6433c9f590b29763

theStack commented on test/functional/rpc_scanblocks.py:
```diff
@@ -94,6 +98,13 @@ def run_test(self):
+
+
         # TODO: after an "accurate" mode for scanblocks is implemented (e.g. PR #26325)
         # check here that it filters out the false-positive
```
These comment lines from the previous commit can be removed now.

aureleoules replied:
Removed thanks

theStack reviewed (approved):
Code-review ACK f0b27d2d8bd76e329ef3b90473b14fe57e079c4e

theStack reviewed (approved):
re-ACK 5fb8f8dc49d00b9e2eb0d6a968783a61449ac683 📔

andrewtoth commented on src/rpc/blockchain.cpp:
```diff
@@ -2244,7 +2280,8 @@ static RPCHelpMan scanblocks()
             RPCArg{"stop_height", RPCArg::Type::NUM, RPCArg::DefaultHint{"chain tip"}, "Height to stop to scan"},
-            RPCArg{"filtertype", RPCArg::Type::STR, RPCArg::Default{BlockFilterTypeName(BlockFilterType::BASIC)}, "The type name of the filter"}
+            RPCArg{"filtertype", RPCArg::Type::STR, RPCArg::Default{BlockFilterTypeName(BlockFilterType::BASIC)}, "The type name of the filter"},
+            RPCArg{"filter_false_positives", RPCArg::Type::BOOL, RPCArg::Default{false}, "Filter false positives. Otherwise they may occur at a rate of 1/M"},
```
Should we document that this RPC requires `blockfilterindex=1` and that it is scanning using the block filters? It doesn't seem like this is noted in the help text above. Then we can say here that block filter scanning has a rate of false positives of 1/M, and that using this option will require Bitcoin Core to read all the blocks from disk and scan them, so it will be more expensive setting this option. Also that this option can fail on a pruned node.

andrewtoth commented on test/functional/rpc_scanblocks.py:
```diff
@@ -71,6 +79,31 @@ def run_test(self):
+        assert(genesis_blockhash in node.scanblocks(
+            "start", [{"desc": f"raw({false_positive_spk.hex()})"}], 0, 0)['relevant_blocks'])
+
+        # check that the filter_false_positives option works
```
Should we also include a test that breaks if the block containing the false positive gets pruned?

aureleoules replied:
Considering that false positives can only be generated in the genesis block and that the genesis block cannot be pruned, I'm not sure how to test this. I will leave this test as a follow-up for a brave soul.

furszy replied:
could hand-craft the first block at the python test suite side.

aureleoules replied:
My assumptions were wrong, I thought the genesis block could not get pruned. Added a test case for this https://github.com/bitcoin/bitcoin/compare/fd0ed48ecd9f0d65acc3199a50b3ca6479aa5a3e..c39234b070a0d2102af835fda77f3a78d4a631d7

aureleoules commented on src/rpc/blockchain.cpp:
```diff
@@ -2244,7 +2280,8 @@ static RPCHelpMan scanblocks()
             RPCArg{"stop_height", RPCArg::Type::NUM, RPCArg::DefaultHint{"chain tip"}, "Height to stop to scan"},
-            RPCArg{"filtertype", RPCArg::Type::STR, RPCArg::Default{BlockFilterTypeName(BlockFilterType::BASIC)}, "The type name of the filter"}
+            RPCArg{"filtertype", RPCArg::Type::STR, RPCArg::Default{BlockFilterTypeName(BlockFilterType::BASIC)}, "The type name of the filter"},
+            RPCArg{"filter_false_positives", RPCArg::Type::BOOL, RPCArg::Default{false}, "Filter false positives. Otherwise they may occur at a rate of 1/M"},
```
I've made the argument `RPCArg::Optional::OMITTED_NAMED_ARG`, is this what you asked? I'm not familiar with RPC args.

luke-jr replied:
No. Listing it in order here makes it a positional argument. Look at RPCs with an "options" argument to see how it should be done.

aureleoules replied:
Ah thanks, I pushed it
https://github.com/bitcoin/bitcoin/compare/c39234b070a0d2102af835fda77f3a78d4a631d7..e94d80dd45c0610c3c40b8a6589c5f7f1d329117

aureleoules commented:
- Added release notes
- Improved RPC documentation
- Made RPC arg `filter_false_positives` "OMITTED_NAMED_ARG"
- Added a check that throws `Filtering false positives is not supported when running with pruning enabled` when pruning and `filter_false_positives` modes are enabled because undo data may be pruned.

aureleoules commented:
Rebased for #26341.

w0xlt reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/26325/commits/fd0ed48ecd9f0d65acc3199a50b3ca6479aa5a3e

aureleoules commented:
I've added a test case in `feature_pruning.py` that checks the `filter_false_positives` mode fails on pruned nodes and the relevant block is pruned, as suggested by @andrewtoth. https://github.com/bitcoin/bitcoin/compare/fd0ed48ecd9f0d65acc3199a50b3ca6479aa5a3e..c39234b070a0d2102af835fda77f3a78d4a631d7

aureleoules commented:
- Moved `filter_false_positives` RPC arg inside `options` to make it optional as suggested by @luke-jr: [diff](https://github.com/bitcoin/bitcoin/compare/c39234b070a0d2102af835fda77f3a78d4a631d7..e94d80dd45c0610c3c40b8a6589c5f7f1d329117)
- Rebased to fix CI error

theStack commented on src/rpc/blockchain.cpp:
```diff
@@ -2240,17 +2241,57 @@ class BlockFiltersScanReserver
+    {
+        LOCK(cs_main);
+        block = GetBlockChecked(blockman, pindex);
+        if (pindex->nHeight > 0) {
+            block_undo = GetUndoChecked(blockman, pindex);
+        }
```
After rebasing on master, excluding the genesis block is not needed anymore, as `GetUndoChecked` takes care of this now and simply returns an empty `CBlockUndo` instance (since #19888, commit 2ca5a496c2f3cbcc63ea15fa05c1658e7f527bbc has been merged):
```suggestion
        block_undo = GetUndoChecked(blockman, pindex);
```

theStack reviewed (commented):
Considering that there is no release out yet that includes the `scanblocks` RPC, this particular release note is not needed I'd say? (We should not forget to add a general one that mentions the introduction of `scanblocks` RPC, but this is unrelated to this PR.)

andrewtoth commented on src/rpc/blockchain.cpp:
```diff
@@ -2265,17 +2266,55 @@ class BlockFiltersScanReserver
+    CBlockUndo block_undo;
+
+    {
+        LOCK(cs_main);
```
`GetBlockChecked` and `GetUndoChecked` both require a lock. https://github.com/bitcoin/bitcoin/pull/26308 removes the lock requirement for these functions. Calling `ReadBlockFromDisk` or `ReadUndoFromDisk` directly doesn't require the lock.

MarcoFalke replied:
Ugh, thanks. Didn't fully grasp the function name

theStack reviewed (approved):
ACK 8da8bee0746a58dae77f790d2695a23f07639830

w0xlt reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/26325/commits/8da8bee0746a58dae77f790d2695a23f07639830

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -2268,17 +2269,49 @@ class BlockFiltersScanReserver
     }
 };
 
+static bool CheckBlockFilterMatches(BlockManager& blockman, const CBlockIndex* pindex, const GCSFilter::ElementSet& needles)
```
```suggestion
static bool CheckBlockFilterMatches(BlockManager& blockman, const CBlockIndex& block, const GCSFilter::ElementSet& needles)
```

Shouldn't this be `&`? (Passing in `nullptr` would be UB)

aureleoules replied:
Thanks, done.

w0xlt reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/26325/commits/c8d5ee3f3f60d50968c7d48f8fdfd678ea2a1fbc

theStack reviewed (approved):
re-ACK c8d5ee3f3f60d50968c7d48f8fdfd678ea2a1fbc

achow101 commented on src/rpc/blockchain.cpp:
```diff
@@ -2268,17 +2269,49 @@ class BlockFiltersScanReserver
+    const CBlock block{GetBlockChecked(blockman, &blockindex)};
+    const CBlockUndo block_undo{GetUndoChecked(blockman, &blockindex)};
+
+    for (const auto& needle : needles) {
+        const auto script{CScript(needle.begin(), needle.end())};
+
+        // Check if any of the outputs match the scriptPubKey
+        for (const auto& tx : block.vtx) {
+            if (std::any_of(tx->vout.cbegin(), tx->vout.cend(), [&](const auto& txout) { return txout.scriptPubKey == script; })) {
+                return true;
+            }
+        }
+
+        // Check if any of the inputs match the scriptPubKey
+        for (const auto& txundo : block_undo.vtxundo) {
+            if (std::any_of(txundo.vprevout.cbegin(), txundo.vprevout.cend(), [&](const auto& coin) { return coin.out.scriptPubKey == script; })) {
+                return true;
+            }
+        }
+    }
```
ISTM this outer loop isn't necessary and could potentially be quite slow. `GCSFilter::ElementSet` is a `std::unordered_set` which has constant time lookup. But with this loop, the lookup becomes linear. It should be much faster to see if each scriptPubKey is contained in `needles`.

```suggestion
// Check if any of the outputs match the scriptPubKey
    for (const auto& tx : block.vtx) {
        if (std::any_of(tx->vout.cbegin(), tx->vout.cend(), [&](const auto& txout) { return needles.count(std::vector<unsigned char>(txout.scriptPubKey.begin(), txout.scriptPubKey.end())) != 0; })) {
            return true;
        }
    }

    // Check if any of the inputs match the scriptPubKey
    for (const auto& txundo : block_undo.vtxundo) {
        if (std::any_of(txundo.vprevout.cbegin(), txundo.vprevout.cend(), [&](const auto& coin) { return needles.count(std::vector<unsigned char>(coin.out.scriptPubKey.begin(), coin.out.scriptPubKey.end())) != 0; })) {
            return true;
        }
    }
```

aureleoules commented:
Thanks @achow101 I added your suggestion.

theStack reviewed (approved):
re-ACK 0b2f0f394248c67a36303c09f7519a9a2d5d29df

I agree that this loop order (as suggested by achow101 in https://github.com/bitcoin/bitcoin/pull/26325#discussion_r1048927397) is the superior approach.

aureleoules commented:
Rebased.

theStack reviewed (approved):
re-ACK 5ca7a7be76f2521dca895daa70949fd11df0844c

achow101 commented:
ACK 5ca7a7be76f2521dca895daa70949fd11df0844c

furszy reviewed (approved):
Code review ACK 5ca7a7be
